File: GFG POTD/032-February-2026/011-equalize-the-towers.java
// https://www.geeksforgeeks.org/problems/equalize-the-towers2804/1/




import java.util.*;

class Solution {
    public int minCost(int[] heights, int[] cost) {
        int low = Arrays.stream(heights).min().getAsInt(); 
        int high = Arrays.stream(heights).max().getAsInt();
        int ans = high;
        while (low < high) {
            int mid = low + (high - low) / 2;
            int v1 = getCost(heights, cost, mid);
            int v2 = getCost(heights, cost, mid + 1);
            if (v1 <= v2) {
                ans = mid;
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return getCost(heights, cost, ans);
    }
    private static int getCost(int[] heights, int[] cost, int tar) {
        int totalCost = 0;
        for (int i = 0; i < heights.length; i ++) {
            totalCost += Math.abs(tar - heights[i]) * cost[i];
        }
        return totalCost;
    }
}

File: GFG POTD/032-February-2026/011-equalize-the-towers.py
# https://www.geeksforgeeks.org/problems/equalize-the-towers2804/1/




class Solution:
    def minCost(self, heights, cost):
        low, high = min(heights), max(heights)
        ans = high
        while low < high:
            mid = low + (high - low) // 2
            v1 = self.getCost(heights, cost, mid)
            v2 = self.getCost(heights, cost, mid + 1)
            if v1 <= v2:
                ans = mid
                high = mid
            else:
                low = mid + 1
        return self.getCost(heights, cost, ans)
        
    def getCost(self, heights, cost, tar):
        totalCost = 0
        for i in range(0, len(heights)):
            totalCost += abs(heights[i] - tar) * cost[i]
        return totalCost